### DISCUSSION TITLE
how to insert link

### DISCUSSION THREAD

**Discussion by solos** [Milkdown/milkdown#1379] [Q&A] [answered]
Hi, 
      This markdown editor is wonderful. I use the 6 version which has a ui menu. The latest version has no menu. I want to insert a link to the editor, but only found the toggleLinkCommand and updateLinkCommand.Hope to get a command like insertImageCommand so that i could insert link with it. 
   I tried the `insert` command, and failed. I found the apply method in website project's linkPlugin, and tried to write a insert link command, but failed. So how to insert a link? thanks.

**Comment by Saul-Mirone** [comment, marked as answer]
Demo: https://milkdown-storybook.vercel.app/?path=/story/components-link-tooltip--default
Code: https://github.com/Milkdown/milkdown/blob/main/e2e/stories/components/link-tooltip.stories.ts#L58

**Reply by solos** [reply]
Thank you very much.

**Comment by solos** [comment]
How to insert link without selecting text first? or is there any way to select some content? 

```ts
  this.editor.action((ctx) => {
        const view = ctx.get(editorViewCtx);
        view.dispatch(view.state.tr.insertText(" "));
        
        // select the inserted text and create selection 
        // don't know how to create selection 
        const selection = 
        view.state.tr.setSelection(selection);
        ctx.get(linkTooltipAPI.key).addLink(selection.from, selection.to);
  });
```

**Reply by Saul-Mirone** [reply]
You will need to know the position you want to apply the selection. And you can apply the selection by `tr.setSelection(TextSelection.create(tr.doc, yourPositionFrom, yourPositionTo))`.

**Reply by solos** [reply]
Cool, I use the link plugin  code and it works. Milkdown is awesome! Thanks.
```
      view.dispatch(view.state.tr.insertText(url));
      const length = url.length;
      const sel = TextSelection.create(view.state.tr.doc, 0, length + 1);
      ctx.update(linkTooltipState.key, (state) => ({
        ...state,
        mode: "edit" as const,
      }));
      view.dispatch(view.state.tr.setSelection(sel));

      const type = linkSchema.type(ctx);

      view.dispatch(
        view.state.tr.addMark(sel.from, sel.to, type.create({ href: url }))
      );

```
